src/slam_karto.cpp

```diff
@@ -291,12 +345,19 @@ SlamKarto::~SlamKarto()
     transform_thread_->join();
     delete transform_thread_;
   }
+  if(vis_thread_)
+  {
+    vis_thread_->join();
+    delete vis_thread_;
+  }
+  if(solver_)
```

Comment:
`delete smart_ptr.get()` is definitely not safe as it deletes the object without decrementing the reference counter.
Call `smart_ptr.reset()` instead. No need for the if clause in that case.

I am surprised that you see this warning as the initialization order looks correct.